### ISSUE TITLE
Twisted defining a _log attribute can cause problems for subclassing users

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#7983]
|[<img alt="hawkowl's avatar" src="https://avatars.githubusercontent.com/u/3307100?s=50" width="50" height="50">](https://github.com/hawkowl)| @hawkowl reported|
|-|-|
|Trac ID|trac#7983|
|Type|defect|
|Created|2015-08-02 00:48:30Z|
|Branch|https://github.com/twisted/twisted/tree/stopsubclassingkids-7983|

In [#7897](https://github.com/twisted/twisted/issues/7897), we ported ```Protocol``` to use the new logging system. This is a good thing, and we need to do it everywhere, but it makes more instance attributes. This leads to namespace conflicts, because people love subclassing.

```
<glyph> hawkowl: I'll just repeat. Put a _log = Logger() at module scope. .Then, do logFor = _log.__get__, then do logFor(self) instead of self.log
<glyph> hawkowl: that is not _heinously_ ugly in the implementation, it gets the job done and we add no new inherited attributes.
```

<details><summary>Searchable metadata</summary>

```
trac-id__7983 7983
type__defect defect
reporter__hawkowl hawkowl
priority__highest highest
milestone__None None
branch__branches_stopsubclassingkids_7983 branches/stopsubclassingkids-7983
branch_author__hawkowl hawkowl
status__closed closed
resolution__fixed fixed
component__core core
keywords__ 
time__1438476510494140 1438476510494140
changetime__1445217440142687 1445217440142687
version__None None
owner__hawkowl hawkowl

```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="hawkowl's avatar" src="https://avatars.githubusercontent.com/u/3307100?s=50" width="50" height="50">](https://github.com/hawkowl)<a name="note_1"></a>|@hawkowl commented|
|-|-|

(In [45336]) Branching to stopsubclassingkids-7983.

**Comment by twisted-trac** [comment]
|[<img alt="hawkowl's avatar" src="https://avatars.githubusercontent.com/u/3307100?s=50" width="50" height="50">](https://github.com/hawkowl)<a name="note_2"></a>|@hawkowl commented|
|-|-|

This fixes Autobahn when this patch is applied. Builders spun, please review.

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_3"></a>|@glyph set owner to @hawkowl|
|-|-|

Thanks for this, hawkowl. 

  1. If you're going to manipulate global state in tests, you have to use `addCleanup`, because a failed test should not result in corrupted state.  So here you would need to do

```
#!py
globalLogPublisher.addObserver(events.append)
self.addCleanup(lambda: globalLogPublisher.remove(events.append))
```
  2. This should really be serving as an example for more of eeshangarg's follow-up work, and I don't feel like `_logFor = lambda _:_log.__get__(_, _.__class__)` is a great example for code clarity we should all aspire to.  That said, this is fixing what is effectively a release-blocking regression, and it at least works, so let's not over-discuss it; please file a follow-up ticket to clean this up and turn it into more of a repeatable pattern.

Please address the above concerns and land.

**Comment by twisted-trac** [comment]
|[<img alt="hawkowl's avatar" src="https://avatars.githubusercontent.com/u/3307100?s=50" width="50" height="50">](https://github.com/hawkowl)<a name="note_4"></a>|@hawkowl set status to `closed`|
|-|-|

(In [45373]) Merge stopsubclassingkids-7983: Don't define _log on Factory in t.i.protocol

Author: hawkowl
Reviewer: glyph
Fixes: [#7983](https://github.com/twisted/twisted/issues/7983)

**Comment by twisted-trac** [comment]
|[<img alt="hawkowl's avatar" src="https://avatars.githubusercontent.com/u/3307100?s=50" width="50" height="50">](https://github.com/hawkowl)<a name="note_5"></a>|@hawkowl commented|
|-|-|

(In [45374]) Merge stopsubclassingkids-7983: Don't define _log on Factory in t.i.protocol

Author: hawkowl
Reviewer: glyph
Fixes: [#7983](https://github.com/twisted/twisted/issues/7983)

**Comment by twisted-trac** [comment]
|[<img alt="hawkowl's avatar" src="https://avatars.githubusercontent.com/u/3307100?s=50" width="50" height="50">](https://github.com/hawkowl)<a name="note_6"></a>|@hawkowl commented|
|-|-|

Ticket retargeted after milestone deleted
